internal/controllers/provisioningrequest_controller.go

```diff
@@ -1184,6 +1192,45 @@ func (t *provisioningRequestReconcilerTask) updateZTPStatus(ctx context.Context,
 	return nil
 }
 
+// updateOCloudNodeClusterId stores the clusterID in the provisionedResources status if it exists.
+func (t *provisioningRequestReconcilerTask) updateOCloudNodeClusterId(ctx context.Context) error {
+	managedCluster := &clusterv1.ManagedCluster{}
+	managedClusterExists, err := utils.DoesK8SResourceExist(
+		ctx, t.client, t.object.Status.ClusterDetails.Name, "", managedCluster)
+	if err != nil {
+		return fmt.Errorf("failed to check if managed cluster exists: %w", err)
+	}
+
+	if managedClusterExists {
+		// If the clusterID label exists, set it in the provisionedResources.
+		clusterID, exists := managedCluster.GetLabels()["clusterID"]
+		if exists {
+			if t.object.Status.ProvisioningStatus.ProvisionedResources == nil {
+				t.object.Status.ProvisioningStatus.ProvisionedResources = &provisioningv1alpha1.ProvisionedResources{}
+			}
+			t.object.Status.ProvisioningStatus.ProvisionedResources.OCloudNodeClusterId = clusterID
+		}
+	}
+	return nil
+}
+
+// finalizeProvisioningIfComplete checks if the provisioning process is completed.
+// If so, it sets the provisioning state to "fulfilled" and updates the provisioned
+// resources in the status.
+func (t *provisioningRequestReconcilerTask) finalizeProvisioningIfComplete(ctx context.Context, allPoliciesCompliant bool) error {
+	if utils.IsClusterProvisionCompleted(t.object) && allPoliciesCompliant {
+		utils.SetProvisioningState(t.object, provisioningv1alpha1.StateFulfilled)
```

Comment:
Proposal - optional: Create a function that would query all the children conditions and determine the state of the parent in one shot. Update the state last, after setting all the children conditions. In my opinion, this would simplify testing and expose  the decision process.
The function would look like this:
```
Calculate the overall state {
// Start with upstream condition and progress to downstream: 

//Configuration applied is first
// "ConfigurationApplied", Missing, Completed, ClusterNotReady, InProgress, TimedOut, OutOfDate

if ConfigurationApplied == Completed, OutOfDate -> set StateCompleted
if ConfigurationApplied == InProgress-> set StateInProgress
if ConfigurationApplied == TimedOut-> set StateTimedOut
if ConfigurationApplied == Missing-> check ClusterProvisioned condition

// ClusterProvisioned is second
// "ClusterProvisioned", Unknown, timedout, Failed, completed

if ClusterProvisioned == Completed, Unknown -> set StateInProgress
if ClusterProvisioned == Failed-> set StateFailed
if ClusterProvisioned == TimedOut-> set StateTimedOut
if ClusterProvisioned == Missing-> check Validation conditions


// HardwareProvisioned is third
// "HardwareProvisioned",Failed, completed 
if HardwareProvisioned == failed  -> set StateFailed
if HardwareProvisioned == Completed -> set StateProgressing

// All Validation conditions are last
// 	"ProvisioningRequestValidated", Failed, completed 
//	"HardwareTemplateRendered",Failed, completed 
//	"ClusterInstanceRendered",Failed, completed ,
//	"ClusterResourcesCreated", Failed, completed ,
//	"ClusterInstanceProcessed", NotApplied, Unkown, completed, failed 

if any == not completed -> set StatePending
if all == Completed or unknown -> set StateProgressing
}
```

Reply:
Thanks for your suggestion. I also considered this approach but decided not choosing it because we would need to ensure it's called before each requeueWithInterval or doNotRequeue in both the `run` and `checkClusterDeployConfigState` functions. I felt that updating it as part of each operation would ensure that it gets updated regardless of whether it's called within `run` or `checkClusterDeployConfigState`.

I'm now updating the process to keep the current 'progressing' status to reflect the ongoing provisioning, even if new changes come in during provisioning that cause failures in preparation (e.g., validation, rendering, or resource creation). Once the process reaches a final state, we will transition to 'failed.' So I now go back considering this option, at least, I'll remove all the state settings from the steps before hardware provisioning. I will see how to make it cleaner.

Reply:
I found out it's simpler to update the state along with each provisioning stage. The check for resource validation and creation error is moved to the end of `checkClusterDeployConfigState` func, ensuring the status reflects to the ongoing provisioning if any.